### PR TITLE
Fix arbitrary audit log ordering in 'get_id_change_date'

### PR DESCRIPTION
**Pull request by ivanb777**
The `get_id_change_date` method retrieves the actual timestamp of modified member attributes using audit logs. However, the ordering of audit logs was not explicitly defined, leading to potential inconsistencies in the returned modification timestamp. This issue could negatively impact the sync process by relying on incorrect timestamps.

This fix ensures that audit logs are properly ordered, guaranteeing the accurate modification timestamp – improving the reliability of the sync process.

### CODE REVIEW

File: app/lib/identity_tijuana/member_sync.rb [hunk 1]
```diff
@@ -24,7 +24,7 @@ def self.export_member(member, sync_id)
         sync(user, member, sync_type, sync_id, :export_member)
       rescue StandardError => e
         Rails.logger.error "[IdentityTijuana::export_member] failed: " \
-                           "#{e.message} member_id=#{member_id}" \
+                           "#{e.message} member_id=#{member.id}" \
```
Comment (michael-gratton) [inline app/lib/identity_tijuana/member_sync.rb:27]: Would have been to to break this out as a seperate commit, but not worth doing now.

Also worth testing this log line by throwing an error in a mocked call to `sync()`, but doing so properly would duplicate some of the test changes in https://github.com/GetUp/identity-tijuana/pull/71 (in fact I may have already fixed that there) - so maybe lets just leave that there.

File: app/lib/identity_tijuana/member_sync.rb [hunk 2]
```diff
@@ -109,15 +109,20 @@ def self.get_id_change_date(member, field_info_type, default_change_date = nil)
       end
       # Iterate through the relevant audit log list, moving backwards in time
       # until we find a change to the field or fields in question.
-      member.send(audit_list_name).where(auditable_type: auditable_type).reverse_each do |audit|
+      # Note that we use `reorder` to remove all prior ordering clauses,
+      # as Audited injectes `order by version asc` when order clause is used.
+      member.send(audit_list_name)
+            .where(auditable_type: auditable_type)
+            .reorder(created_at: :desc)
+            .each do |audit|
```
Comment (michael-gratton) [inline app/lib/identity_tijuana/member_sync.rb:117]: Can you walk me through why the versioning is not reliable but the created_at is?

File: app/lib/identity_tijuana/member_sync.rb [hunk 3]
```diff
@@ -109,15 +109,20 @@ def self.get_id_change_date(member, field_info_type, default_change_date = nil)
...
         # Check ancillary matching field, where required.
         if !ancillary_match_value.nil?
           case auditable_type
           when 'PhoneNumber'
             phone_number_type =
               audit.audited_changes['phone_type'] ||
               PhoneNumber.find_by(id: audit.auditable_id).try(:phone_type)
-            next unless phone_number_type == ancillary_match_value
+            next unless phone_number_type.to_sym == ancillary_match_value
```
Comment (michael-gratton) [inline app/lib/identity_tijuana/member_sync.rb:125]: Is `ancillary_match_value` always a symbol when not nil? From the places its set above, it looks like it could also sometimes be a string?

File: spec/lib/identity_tijuana_member_sync_spec.rb [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+        "geography.area_lookup.track_area_probabilities"
+      ) { false }
+    end
+
+    context 'with multiple audit logs entries' do
+      it 'correctly identifies the change date for name' do
+        member = FactoryBot.create(:member)
+        member.update!(first_name: 'NewName')
+
+        sleep(0.2)
```
Comment (michael-gratton) [inline spec/lib/identity_tijuana_member_sync_spec.rb:36]: Prefer using the [time helpers](https://guides.rubyonrails.org/testing.html#testing-time-dependent-code) instead of sleeping since it gives you more control, and doesn't slow down the unit tests.

**Review by michael-gratton — CHANGES_REQUESTED** [review]

File: app/lib/identity_tijuana/member_sync.rb (hunk 3 above)
Comment (michael-gratton) [reply app/lib/identity_tijuana/member_sync.rb:125]: NB I don't know if Ruby returns true or false for `"foo".to_sym == "foo"`, but if false maybe this line isn't needed at all?

File: app/lib/identity_tijuana/member_sync.rb (hunk 1 above)
Comment (ivanb777) [reply app/lib/identity_tijuana/member_sync.rb:27]: yeah, I saw that you've fixed `member_id`, I'll remove the fix from this PR – it won't break the tests

File: app/lib/identity_tijuana/member_sync.rb (hunk 3 above)
Comment (ivanb777) [reply app/lib/identity_tijuana/member_sync.rb:125]: > Is ancillary_match_value always a symbol when not nil?

No, it is only a symbol when `auditable_type` is `PhoneNumber` -> `ancillary_match_value = field_info_type` with call eg. `get_id_change_date(member, :mobile, id_mobile&.updated_at || member&.updated_at)`.
Otherwise, it's integer eg. `ancillary_match_value = Settings.tijuana.email_subscription_id`

> NB I don't know if Ruby returns true or false for "foo".to_sym == "foo", but if false maybe this line isn't needed at all?

`"foo".to_sym == "foo"` is `false`, but with `PhoneNumber` case we want to test `"foo".to_sym == :foo`, as in
`"mobile".to_sym == :mobile`.

File: app/lib/identity_tijuana/member_sync.rb (hunk 2 above)
Comment (ivanb777) [reply app/lib/identity_tijuana/member_sync.rb:117]: > Can you walk me through why the versioning is not reliable but the created_at is?

`member.send(audit_list_name).where(auditable_type: auditable_type).reverse_each` query will not apply any ordering, and the resulting set is arbitrarily ordered.

The `ORDER BY version ASC` is injected as the primary order clause only when `order` method is called. In essence, yes we can call `order(version: :desc)` and we'll get the same ordering.

I've opted for `reorder` with `created_at` as it is indexed column, as opposed to `version`, but it also made more sense to keep aligned with `moving backwards in time` in the iteration through the relevant audit logs.

File: app/lib/identity_tijuana/member_sync.rb (hunk 1 above)
Comment (ivanb777) [reply app/lib/identity_tijuana/member_sync.rb:27]: Moved this to a [separate commit](https://github.com/GetUp/identity-tijuana/pull/73/commits/c440d13a2cc98c72d6578c076e572faefbbc159e)

File: spec/lib/identity_tijuana_member_sync_spec.rb (hunk 4 above)
Comment (ivanb777) [reply spec/lib/identity_tijuana_member_sync_spec.rb]: /nice – changed to `travel_to`

File: app/lib/identity_tijuana/member_sync.rb (hunk 2 above)
Comment (michael-gratton) [reply app/lib/identity_tijuana/member_sync.rb:117]: Got it, thanks. Can you update the comment to say something similar?

It's clear from the code and the `reorder` method's API docs that the reordering is happening and what its effects are, but it's not clear why that is needed.

**Review by michael-gratton — CHANGES_REQUESTED** [review]

File: app/lib/identity_tijuana/member_sync.rb (hunk 3 above)
Comment (michael-gratton) [reply app/lib/identity_tijuana/member_sync.rb:125]: Again, I'm not sure if `1 == "1"` will eval to true or false in ruby, but I hope the latter. In which case the tests may well be behaving differently to runtime when integrated with Id.

Comment (ivanb777) [reply app/lib/identity_tijuana/member_sync.rb:125]: ahh, I totally missed that! 
well, yes `1 == "1"` is `false`, we should really cast these `audit_subscription_id` and `ancillary_match_value` to int when the `auditable_type` is `MemberSubscription`

**Review by michael-gratton — APPROVED** [review]
Let's get this deployed to staging!
